File: 36_graphs/36_12_days_until_all_infected.py
from collections import deque

def days_until_all_infected(graph: list[list[int]], infected: list[int])->int:
  queue = deque()
  infected_map = {}
  # Init
  for node in infected:
    infected_map[node] = 0
    queue.append((node, 0)) # node, days

  while len(queue) > 0:
    node, days = queue.popleft()
    for nei in graph[node]:
      if nei not in infected_map:
        infected_map[nei] = days + 1
        queue.append((nei, days + 1))

  max_days = 0
  for days in infected_map.values():
    if days > max_days:
      max_days = days
  
  return max_days

if __name__ == "__main__":
  # Example1
  graph = [
    [1, 2],
    [0, 2],
    [0, 1, 3],
    [2]
  ]
  infected = [0]
  print(days_until_all_infected(graph, infected)) # Output: 2

  # Example2
  graph = [
    [1],
    [0, 2],
    [1, 3],
    [2, 4],
    [3]
  ]
  infected = [0, 4]
  print(days_until_all_infected(graph, infected)) # Output: 2

  # Example3
  graph = [
    [1, 2],
    [0, 3],
    [0, 3],
    [1, 2]
  ]
  infected = [0, 3]
  print(days_until_all_infected(graph, infected)) # Output: 1

File: 36_graphs/36_12_days_until_all_infected.cpp
#include <vector>
#include <queue>
#include <unordered_map>
#include <iostream>

int days_until_all_infected(std::vector<std::vector<int>>& graph, std::vector<int>& infected) {
  std::queue<std::pair<int, int>> queue;
  std::unordered_map<int, int> infected_map;

  for (int node : infected) {
    infected_map[node] = 0;
    queue.push({node, 0});
  }
  
  while(!queue.empty()) {
    std::pair<int, int> current = queue.front();
    queue.pop();
    int node = current.first;
    int days = current.second;

    for (int nei : graph[node]) {
      if (infected_map.find(nei) == infected_map.end()) {
        infected_map[nei] = days + 1;
        queue.push({nei, days + 1});
      }
    }
  }

  int max_days = 0;
  for (const auto& pair : infected_map) {
    if (pair.second > max_days) {
      max_days = pair.second;
    }
  }

  return max_days;
}

int main() {
    // Example1
    std::vector<std::vector<int>> graph1 = {
        {1, 2},
        {0, 2},
        {0, 1, 3},
        {2}
    };
    std::vector<int> infected1 = {0};
    std::cout << days_until_all_infected(graph1, infected1) << std::endl; // Output: 2

    // Example2
    std::vector<std::vector<int>> graph2 = {
        {1},
        {0, 2},
        {1, 3},
        {2, 4},
        {3}
    };
    std::vector<int> infected2 = {0, 4};
    std::cout << days_until_all_infected(graph2, infected2) << std::endl; // Output: 2

    // Example3
    std::vector<std::vector<int>> graph3 = {
        {1, 2},
        {0, 3},
        {0, 3},
        {1, 2}
    };
    std::vector<int> infected3 = {0, 3};
    std::cout << days_until_all_infected(graph3, infected3) << std::endl; // Output: 1

    return 0;
}
